src/inbox/mod_inbox_odbc.erl

```diff
@@ -0,0 +1,165 @@
+        {selected, []} ->
+            [];
+        {selected, Res} ->
+            [decode_row(LServer, R) || R <- Res]
+    end.
+
+
+-spec get_inbox_rdbms(LUser :: jid:luser(), Server :: jid:lserver()) -> query_result().
+get_inbox_rdbms(LUser, Server) ->
+    mongoose_rdbms:sql_query(
+        Server,
+        ["select remote_bare_jid, content, unread_count from inbox "
+        "where luser=", ?ESC(LUser), " and lserver=", ?ESC(Server), ";"]).
+
+-spec set_inbox(Username, Server, ToBareJid, Content, Count, MsgId) -> inbox_write_res() when
+                Username :: jid:luser(),
+                Server :: jid:lserver(),
+                ToBareJid :: binary(),
+                Content :: binary(),
+                Count :: binary(),
+                MsgId :: binary().
+set_inbox(Username, Server, ToBareJid, Content, Count, MsgId) ->
+    LUsername = jid:nodeprep(Username),
+    LServer = jid:nameprep(Server),
+    LToBareJid = jid:nameprep(ToBareJid),
+    BackendModule = odbc_specific_backend(Server),
+    Res = BackendModule:set_inbox(LUsername, LServer, LToBareJid, Content, Count, MsgId),
+    ok = check_result(Res, 1).
+
+-spec remove_inbox(User :: binary(),
+    Server :: binary(),
+    ToBareJid :: binary()) -> ok.
+remove_inbox(Username, Server, ToBareJid) ->
+    LUsername = jid:nodeprep(Username),
+    LServer = jid:nameprep(Server),
+    LToBareJid = jid:nameprep(ToBareJid),
+    Res = remove_inbox_rdbms(LUsername, LServer, LToBareJid),
+    check_result(Res).
+
+-spec remove_inbox_rdbms(Username :: jid:luser(),
+                         Server :: jid:lserver(),
+                         ToBareJid :: binary()) -> query_result().
+remove_inbox_rdbms(Username, Server, ToBareJid) ->
+    mongoose_rdbms:sql_query(Server, ["delete from inbox where luser=",
+        ?ESC(Username), " and lserver=", ?ESC(Server),
+        " and remote_bare_jid=",
+        ?ESC(ToBareJid), ";"]).
+
+-spec set_inbox_incr_unread(Username :: binary(),
+                            Server :: binary(),
+                            ToBareJid :: binary(),
+                            Content :: binary(),
+                            MsgId :: binary()) -> ok.
+set_inbox_incr_unread(Username, Server, ToBareJid, Content, MsgId) ->
+    LUsername = jid:nodeprep(Username),
+    LServer = jid:nameprep(Server),
+    LToBareJid = jid:nameprep(ToBareJid),
+    BackendModule = odbc_specific_backend(Server),
+    Res = BackendModule:set_inbox_incr_unread(LUsername, LServer, LToBareJid, Content, MsgId),
+    %% psql will always return {updated, 1} but mysql will return {updated, 2} if it overwrites the row
+    check_result(Res,[1,2]).
+
+-spec reset_unread(User :: binary(),
+                   Server :: binary(),
+                   BareJid :: binary(),
+                   MsgId :: binary()) -> ok.
+reset_unread(Username, Server, ToBareJid, MsgId) ->
+    LUsername = jid:nodeprep(Username),
+    LServer = jid:nameprep(Server),
+    LToBareJid = jid:nameprep(ToBareJid),
```

Comment:
while the current version "works", I would use 

```erlang
jid_prep(Jid) ->
     jid:to_binary(jid:to_lower(jid:from_binary(Jid))).
```

Reply:
which can go to jid module.

Reply:
also, prep functions should go into the main module, not into the backend module!
backend modules should have bare minimum of logic.